### ISSUE TITLE
CI: GHA ubuntu-18.04 image no longer works

### ISSUE THREAD

**Issue by ojwb** [swig/swig#2537]
The jobs using the ubuntu-18.04 image had started to sit un-run for a day and then fail, so I commented them all out in f030df8786bc3e707bee9079d093368de3d8bc57 since CI jobs which always fail just get in the way.

Since then, I've resurrected the scilab 6.0 job in 5c165f5c, and all the ruby jobs except ruby 1.9 in 1ec05d04.

I've also tested updating the rest to 20.04, but they all fail, so currently these jobs are still disabled:

* ~~Node 6, 8, 10~~ Addressed by documenting Node 12 as the lowest supported version
* ~~Python 3.3, 3.4~~ Addressed by documenting we're no longer testing with these versions
* ~~Ruby 1.9~~ Addressed by documenting Ruby 2.0 as the lowest supported version
* ~~Ocaml (only job)~~ We now have Ocaml CI testing again

~~I also notice the only SWIG/D job was already disabled as it apparently fails on both 18.04 and 20.04.  That means we have had no CI coverage for SWIG/D since at least 2021-09-30 when we switched to GHA.  Maybe D should be moved to "experimental" status in SWIG 4.2.0 if nobody steps up to sort out CI coverage by then.~~ CI testing for D was restored in May 2023.

@vadz suggested we could use a docker container to get 18.04 in https://github.com/swig/swig/commit/f030df8786bc3e707bee9079d093368de3d8bc57#commitcomment-109388830 but I think that's beyond my skills and enthusiasm to do.  I also wonder if it's worth the effort as Ubuntu 18.04 is out of support 2023-05-31 (there's extended support after that, but AIUI that's a commercial offering that requires paying Canonical).

(I also continue to think that given our limited development resources we put too much effort into trying to support ancient versions of target languages which comes at the cost of getting other things many users would find useful done.)

**Comment by ojwb** [comment]
The ocaml testsuite works fine for me locally (on Debian unstable; there's some warnings from the examples, but they also work) however the CI job fails on ubuntu-20.04 with:

```
 checking ocaml test-suite
File "swig.ml", line 1:
Error: Could not find the .cmi file for interface swig.mli.
```

For me, `swig.cmi` gets built fine:

```
env SWIG_LIB=../../../Lib ../../../swig -ocaml -co swig.mli
ocamlc -c swig.mli
```

I've not managed to work out what's going wrong there - configure seems to successfully probe the same tools for ocaml in CI as for me.

On ubuntu-22.04 it fails in a different way, but working on 20.04 would be fine.

@ZackerySpytz  I wonder if the problem would be obvious to someone who knows ocaml?

**Comment by ojwb** [comment]
I hit a problem with javascript which only manifested in CI and turned out to be due to srcdir!=builddir (whereas locally I'm building in the source tree) - maybe that's the case here too?

**Comment by wsfulton** [comment]
HI @ojwb. Thanks for getting GHA green again. I've been away for a while, but will take a look to see if I can resurrect the disabled builds.

**Comment by ojwb** [comment]
Related, I noticed https://github.com/actions/runner-images/issues/7401 which says:

> Python 2.7 will be removed from the images' tool cache.
> [...]
> The image rollout will start on May 15, 2023 and take 3-4 days.

It looks like we just install the `python-dev` package from the Ubuntu repo to get Python 2 so I suspect this won't affect us, but should Python 2 builds start failing next week this is probably why.

**Comment by ojwb** [comment]
> I hit a problem with javascript which only manifested in CI and turned out to be due to srcdir!=builddir (whereas locally I'm building in the source tree) - maybe that's the case here too?

I tried building in `build/build` (like our GHA CI does) and `make check-ocaml-test-suite` passes for me so this isn't the problem.

**Comment by ojwb** [comment]
I was poking the Ruby builds because they've been failing a lot recently (conclusion: probably just flakiness somewhere between github and launchpad's PPAs) and while I was there I tried re-enabling the Ruby 1.9 build.  Sadly it still fails, apparently because rubygems no longer works with such an old version, so I've updated the manual to say "_SWIG 4.2 is known to work with Ruby versions 2.0 and later_" since it's now been 6 months without any CI testing of 1.9.  Upstream end of security support for Ruby 1.9 was Feb 2015.

I've also added a Ruby 3.3 build as Ruby 3.3.0-preview2 is out.  The testsuite already passes with it without any issues.

**Comment by ojwb** [comment]
We haven't made further progress on the missing builds and realistically I don't see us doing so, and not having CI for older versions of a target language means sooner or later it'll stop working and we just won't know.  We require CI for new target languages so setting the supported version cut-off based on what we can test in CI makes a lot of sense and I suggest we should address this for at least some affected target languages by raising the lowest supported versions for SWIG 4.3.

E.g. [according to wikipedia](https://en.wikipedia.org/wiki/Node.js#Releases), node 6.x, 8.x and 10.x end of support dates were 2019-04-30, 2019-12-31 and 2021-04-30 respectively (the oldest version still supported upstream is 18.x) so 6.x and 8.x are already over 4 years past EOL and 10.x will be over 3 years past EOL by the time SWIG 4.3 is released.

**Comment by ojwb** [comment]
We resolved ocaml (the issue hit above turned out to be due to parallelism not working for our ocaml testsuite - see #2833).

For now I've noted the oldest Node and Python3 versions we regularly test with in the manual (12 and 3.5 respectively).  I think we should seriously consider aligning the minimum supported versions with what we're actually able to CI test (or if anyone strongly disagrees with doing so, the onus should be on them to sort out CI testing for these older versions, perhaps via running in a docker container, or finding a way to install the older version on a GHA-supported Ubuntu version).  Adding to the 4.3 milestone.

**Comment by ojwb** [comment]
I've adjusted the wording in the manual to better align what we claim to support and what we actually test with.

One gap I noticed is that we don't actually seem to have any CI testing for Javascript V8 directly.  Much of the V8-specific code is also used by Node, but now we only test Node 12 and later than means we only have any testing of V8 7.4 and later.  That's notable because it means we're no longer exercising one side of any of the conditionals on the V8 version:

```
Lib/javascript/v8/javascriptcode.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if SWIG_V8_VERSION < 0x0608
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
Lib/javascript/v8/javascriptrun.swg:#if (SWIG_V8_VERSION < 0x0704)
```

For now I've left in the claim that V8 5.0 or later is supported as we don't have CI to back up claims of support for any V8 version range, but I think it would be more realistic to say 7.4 or later, and to drop conditionalised code for anything older.  I've not seem any bug reports against `-v8` that I can recall, and it seems likely to be a very niche thing if used at all.

**Comment by ojwb** [comment]
> For now I've left in the claim that V8 5.0 or later is supported as we don't have CI to back up claims of support for any V8 version range, but I think it would be more realistic to say 7.4 or later, and to drop conditionalised code for anything older. I've not seem any bug reports against `-v8` that I can recall, and it seems likely to be a very niche thing if used at all.

I've had a go at reinstating v8 CI jobs, and it seems pretty bit-rotten.  I've fixed the stuff I easily can, but `Tools/javascript/v8_shell.cxx` doesn't compile with the v8 versions we can easily test with in CI so we can't actually run any tests.  `git log` shows it hasn't been touched in 10 years aside from cosmetic tweaks and removing support for older v8 versions we dropped support for, so it's not a huge surprise it needs updating.  I had a poke but for example I have no idea how to replace the use of `SetHiddenValue()` and `GetHiddenValue()` which no longer exist.  I just have no familiarity with the v8 API so I'm lost at this point.

The full compilation failure is at:

https://github.com/swig/swig/actions/runs/10916679995/job/30298560925#step:8:88

I'll try to tidy up what I've done so far and push that at least.

**Comment by ojwb** [comment]
> I had a poke but for example I have no idea how to replace the use of `SetHiddenValue()` and `GetHiddenValue()` which no longer exist. I just have no familiarity with the v8 API so I'm lost at this point.

I realised that there's only one instance of `V8Shell` at a time so we can just replace this with a global variable (which is less clean, but this is only part of our test harness and less clean seems preferable to completely non-functional with any recent v8 version).

I hacked away the remaining pile of errors and I think I now have the examples and tests passing.  I bodged one testcase to pass for now (it was added since we lost V8 CI so never got tested there before) which I need to return to and fix in a cleaner way, but otherwise it seems OK.

**Comment by ojwb** [comment]
I've cleaned up and merged the v8 fixes.  It's not a fun job as the v8 API seems to evolve fairly rapidly and I didn't find a clear list of what changed when and how to update code, so I'm probably not going to step in and do this going forwards, but at least it now works and we have CI coverage so it won't get accidentally broken by future changes to SWIG, and we'll know if v8 API changes break it again.

I think everything here is now addressed so closing.
